cli/src/git_util.rs

```diff
@@ -139,21 +140,93 @@ fn get_ssh_keys(_username: &str) -> Vec<PathBuf> {
 }
 
 pub fn with_remote_git_callbacks<T>(
-    ui: &mut Ui,
+    ui: &Ui,
+    print_sideband_messages: bool,
```

Comment:
ISTR an earlier version of this PR doing this but, wasn't the decision to print things actually based on the return value of e.g. `push_refs`? I assume you refactored it this way for a good reason. However, it would be nice if instead we could *collect* sideband messages and decide to print them at the call site, instead.

If you check out the Gerrit PR, this would be convenient because we typically run `push_refs` potentially multiple times, so I'd like to be able to "bundle up" all the remote messages from these multiple calls, and print them at my own leisure.

Perhaps instead, `with_remote_callbacks` could take an `Option(&mut Vec<String>)` or something instead of a bool, so that the caller can collect and do what they want with the messages.

I don't think this is a blocker, BTW. I'll happily look into refactoring this myself if after the fact, if necessary.

Reply:
This is related to this comment:

> However, It also seems forges like GitHub makes use of carriage returns to display a more interactive push message (eg. `Resolving deltas: (0/5)\r...\rResolving deltas: 100% (5/5), done.`) to display progress as content is being pushed before finally showing the "Create a pull request" message (or other messages).

So we print lines as we get them (not waiting for operation to finish) to support progress bars and other potential terminal animations 🤷

p.s. omg now I want to do the `curl parrot.live` thing except you're pushing to a fake git remote lol

Reply:
Yeah, I think the best way would be for `with_git_remote_callbacks` to accept a callback which does whatever it wants with each message as it comes in, however I'm not too sure right now how to get it working with Rust.

Reply:
I've tried out https://github.com/bnjmnt4n/jj/commit/debf5d04cfd7dc71d96bb2c0737a988d72d560ab which uses a trait object to store the callback? Not sure if it follows Rust conventions, but I couldn't figure out the right types to use to get pass a closure within an `Option`. Maybe someone better versed in Rust can help with this?

Reply:
well fyi another way is to use a clouse (something like `Box<dyn FnMut(&Ui, &[u8])>`) instead and have
```rust
let mut scratch = ...;
let display_prefix = ...;
let suffix = ...;

let sideband_progress_callback = Box::new(|ui, progress_message| {
    // ... DefaultGitSidebandProgressCallback::on_progress inlined
});
```

But your version is equally good, maybe even better :)

Reply:
After a bit of trial and error, I figured out how to use the closure as well: https://github.com/bnjmnt4n/jj/commit/6ecd6e7e9a54b464b0037f2f622b74d8ceefc931. I was thinking the trait object might be a little too verbose.

Reply:
I've squashed the above commit into my latest push, where I used the closure instead:

```rust
let mut scratch: Vec<u8> = Vec::new();
let mut write_message = get_git_sideband_progress_message_writer();
let mut sideband_progress_callback = |ui: &Ui, progress_message: &[u8]| {
    write_message(ui, &mut scratch, progress_message);
};
with_remote_git_callbacks(ui, Some(&mut sideband_progress_callback), fn);
```

@thoughtpolice I think it would suffice to clone each `&[u8]` and store them in a separate vector, and then write all the messages out at the end.

Reply:
Thanks for looking into it! This looks like it will work well; if I have any changes needed for the Gerrit PR I can do them in a follow up myself.